File: docs/_plugins/block-fields.mjs
// MyST plugin: `block_fields` directive.
//
// Replaces the old Sphinx `block_fields` directive
// (common/python/sphinx_block_fields_directive.py). Renders a block's
// Name / Type / Description table from its `*.block.ini`.
//
// Parsing is delegated to the standalone `common/python/block_fields.py` CLI
// (shelled out via `--format json`), so the docs use the same canonical
// configparser-based reading as the firmware build rather than a second,
// JS-side parser. Set the PYTHON env var to choose the interpreter.

import path from 'node:path';
import { execFileSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';

// Directive args are written relative to the repo root (e.g.
// `modules/counter/counter.block.ini`); this plugin lives in docs/_plugins.
const REPO_ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', '..');
const PYTHON = process.env.PYTHON || 'python3';
const RENDERER = path.join(REPO_ROOT, 'common', 'python', 'block_fields.py');

function textCell(value, header = false) {
  const cell = { type: 'tableCell', children: [{ type: 'text', value: String(value) }] };
  if (header) cell.header = true;
  return cell;
}

function row(values, header = false) {
  return { type: 'tableRow', children: values.map((v) => textCell(v, header)) };
}

const blockFields = {
  name: 'block_fields',
  doc: "Render a block's fields from its *.block.ini as a Name/Type/Description table.",
  arg: { type: String, required: true, doc: 'Path to the .block.ini, relative to the repo root.' },
  run(data, vfile) {
    const iniPath = path.resolve(REPO_ROOT, data.arg);
    let fields;
    try {
      const out = execFileSync(PYTHON, [RENDERER, iniPath, '--format', 'json'], { encoding: 'utf8' });
      fields = JSON.parse(out);
    } catch (err) {
      vfile.message(`block_fields: failed to render '${data.arg}': ${err.message}`);
      return [];
    }

    const rows = [row(['Name', 'Type', 'Description'], true)];
    for (const f of fields) rows.push(row([f.name, f.type, f.description]));
    return [{ type: 'table', children: rows }];
  },
};

const plugin = { name: 'PandABlocks block_fields', directives: [blockFields] };
export default plugin;

File: common/python/block_fields.py
#!/usr/bin/env python
"""Render a block's fields from its ``<block>.block.ini``.

Standalone (no relative imports) so it can be run directly as a command-line
tool *and* be shelled out to from the MyST ``block_fields`` directive
(docs/_plugins/block-fields.mjs). Parsing uses :mod:`configparser`, the
canonical reader for these files, so the docs match the firmware build.

Examples::

    python common/python/block_fields.py modules/counter/counter.block.ini
    python common/python/block_fields.py modules/counter/counter.block.ini --format json
"""
import argparse
import configparser
import json
import sys


def read_fields(path):
    """Return ``[{name, type, description}, ...]`` for each field in the ini.

    The block-level ``[.]`` section is skipped; everything else is a field.
    """
    ini = configparser.ConfigParser()
    if not ini.read(path):
        raise SystemExit("block_fields: cannot read ini file %r" % (path,))
    fields = []
    for section in ini.sections():
        if section == ".":
            continue
        fields.append({
            "name": section,
            "type": ini.get(section, "type", fallback=""),
            "description": ini.get(section, "description", fallback=""),
        })
    return fields


def format_table(fields):
    """Format fields as a simple aligned text table for command-line use."""
    rows = [("Name", "Type", "Description")]
    rows += [(f["name"], f["type"], f["description"]) for f in fields]
    w0 = max(len(r[0]) for r in rows)
    w1 = max(len(r[1]) for r in rows)
    lines = []
    for i, (name, typ, desc) in enumerate(rows):
        lines.append("%-*s  %-*s  %s" % (w0, name, w1, typ, desc))
        if i == 0:
            lines.append("%s  %s  %s" % ("-" * w0, "-" * w1, "-" * 11))
    return "\n".join(lines)


def main(argv=None):
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("ini", help="path to the <block>.block.ini file")
    parser.add_argument(
        "--format", choices=["table", "json"], default="table",
        help="output format: human-readable table (default) or json")
    args = parser.parse_args(argv)

    fields = read_fields(args.ini)
    if args.format == "json":
        json.dump(fields, sys.stdout)
    else:
        print(format_table(fields))
    return 0


if __name__ == "__main__":
    sys.exit(main())
